presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -583,6 +663,139 @@ public Void visitCall(CallExpression call, Context context)
+                        LambdaDefinitionExpression lambda = (LambdaDefinitionExpression) callExpression.getArguments().get(arrayFunctionDescriptor.getFunctionArgumentIndex());
+
+                        List<Set<Subfield>> accessedSubfieldsList = arrayFunctionDescriptor.getArrayArgumentIndexes().stream()
+                                .map(ignored -> new HashSet<>(accessedSubfields))
+                                .collect(Collectors.toList());
+                        Context subContext = new Context();
+                        lambda.getBody().accept(this, subContext);
+                        for (int lambdaArgumentIndex : arrayFunctionDescriptor.getLambdaArgumentToArrayArgumentIndexMap().keySet()) {
+                            int arrayArgumentIndex = arrayFunctionDescriptor.getLambdaArgumentToArrayArgumentIndexMap().get(lambdaArgumentIndex);
+                            String root = lambda.getArguments().get(lambdaArgumentIndex);
+                            accessedSubfieldsList.get(arrayArgumentIndex)
+                                    .addAll(subContext.subfields.stream().filter(x -> x.getRootName().equals(root)).collect(Collectors.toSet()));
+                        }
+                        for (Integer arrayArgumentIndex : arrayFunctionDescriptor.getArrayArgumentIndexes()) {
+                            if (!recursivelyExtractSubfieldsFromArrayLambdas(
+                                    rootCall,
+                                    callExpression.getArguments().get(arrayArgumentIndex),
+                                    context,
+                                    accessedSubfieldsList.get(arrayArgumentIndex),
+                                    prefixPath)) {
+                                return false;
+                            }
+                        }
+                        return true;
+                    }
+
+                    for (Integer arrayArgumentIndex : arrayFunctionDescriptor.getArrayArgumentIndexes()) {
+                        if (!recursivelyExtractSubfieldsFromArrayLambdas(rootCall, callExpression.getArguments().get(arrayArgumentIndex), context, accessedSubfields, prefixPath)) {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+                else if ((expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm() == IF)) {
+                    for (RowExpression argument : ((SpecialFormExpression) expression).getArguments()) {
+                        if (!recursivelyExtractSubfieldsFromArrayLambdas(rootCall, argument, context, accessedSubfields, prefixPath)) {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+
+                Optional<Subfield> subfield = toSubfield(expression, functionResolution, expressionOptimizer, connectorSession, metadata);
```

Comment:
what does this case cover?

Reply:
The function `MAP_TOP_N_VALUES(m, 10, (x, y) -> IF(x.d1 < y.d1, -1, IF(x.d1 = y.d1, 0, 1)))` is a syntactic sugar to the expression below. 

```
SLICE(
    CONCAT(
        REVERSE(
            ARRAY_SORT(
                IF(
                    NONE_MATCH(MAP_VALUES(m), (x$lambda_6) -> IS_NULL(x$lambda_6)),
                    MAP_VALUES(m),
                    FILTER(MAP_VALUES(m), (x$lambda_7) -> NOT(IS_NULL(x$lambda_7)))
                ),
                (expr, expr_0) -> IF(
                        LESS_THAN(DEREFERENCE(expr, 0), DEREFERENCE(expr_0, 0)),
                        -1,
                        IF(
                            EQUAL(DEREFERENCE(expr, 0), DEREFERENCE(expr_0, 0)),
                            0,
                            1
                        )
                    )
            )
        ),
        FILTER(MAP_VALUES(m), (x$lambda) -> IS_NULL(x$lambda))
    ),
    1,
    10
)
```